sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala

```diff
@@ -40,14 +40,19 @@ class StateSchemaCompatibilityChecker(
       logDebug(s"Schema file for provider $providerId exists. Comparing with provided schema.")
       val (storedKeySchema, storedValueSchema) = readSchemaFile()
 
-      def fieldToType: StructField => (DataType, Boolean) = f => (f.dataType, f.nullable)
-
-      def typesEq(schema1: StructType, schema2: StructType): Boolean = {
-        (schema1.length == schema2.length) && schema1.map(fieldToType) == schema2.map(fieldToType)
+      def fieldCompatible(fieldOld: StructField, fieldNew: StructField): Boolean = {
+        // compatibility for nullable
+        // - same: OK
+        // - non-nullable -> nullable: OK
+        // - nullable -> non-nullable: Not compatible
+        (fieldOld.dataType == fieldNew.dataType) &&
+          ((fieldOld.nullable == fieldNew.nullable) ||
+            (!fieldOld.nullable && fieldNew.nullable))
       }
 
-      def namesEq(schema1: StructType, schema2: StructType): Boolean = {
-        schema1.fieldNames.sameElements(schema2.fieldNames)
+      def schemaCompatible(schemaOld: StructType, schemaNew: StructType): Boolean = {
+        (schemaOld.length == schemaNew.length) &&
```

Comment:
Just for my own understanding what will happen when column dropped in `schemaNew`?

Reply:
Undefined behavior, as state would be stored in file as unsafe byte array, and we just rely on new schema to parse it. It might be fine while reading if all column(s) is(are) dropped from rightmost, but the some of information in row (like numFields) would be incorrect so not sure which operation refers it and finally make query crash. If column(s) is(are) dropped from other spots, query would be crashed sooner.

Reply:
Thanks for explaining.